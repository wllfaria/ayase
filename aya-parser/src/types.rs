#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum Operator {
    Add,
    Sub,
    Mul,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Atom<'parser> {
    HexLiteral(&'parser str),
    Address(&'parser str),
    Register(&'parser str),
    Var(&'parser str),
    Operator(Operator),
    BinaryOp {
        lhs: Box<Atom<'parser>>,
        operator: Operator,
        rhs: Box<Atom<'parser>>,
    },
}

#[derive(Debug, PartialEq, Eq)]
pub enum Instruction<'parser> {
    MovLitReg(Atom<'parser>, Atom<'parser>),
    MovRegReg(Atom<'parser>, Atom<'parser>),
    MovRegMem(Atom<'parser>, Atom<'parser>),
    MovMemReg(Atom<'parser>, Atom<'parser>),
    MovLitMem(Atom<'parser>, Atom<'parser>),
    MovRegPtrReg(Atom<'parser>, Atom<'parser>),
}
