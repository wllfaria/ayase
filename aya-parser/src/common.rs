use nom::branch::alt;
use nom::bytes::complete::{tag, tag_no_case};
use nom::character::complete::{alpha1, alphanumeric1, anychar, char, hex_digit1, multispace0, one_of};
use nom::combinator::{map, peek as nom_peek, recognize};
use nom::multi::many0_count;
use nom::sequence::{pair, terminated};
use nom::IResult;

use crate::types::{Atom, Operator};

pub mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 1;
    pub const MUL: u8 = 2;
}

pub fn get_precedence(operator: Operator) -> u8 {
    match operator {
        Operator::Add | Operator::Sub => precedences::SUM,
        Operator::Mul => precedences::MUL,
    }
}

pub fn is_terminator(ch: char) -> bool {
    matches!(ch, ']' | ')')
}

pub fn peek(input: &str) -> IResult<&str, char> {
    nom_peek(anychar)(input)
}

pub fn general_register(input: &str) -> IResult<&str, &str> {
    recognize(pair(tag_no_case("r"), one_of("12345678")))(input)
}

pub fn register(input: &str) -> IResult<&str, Atom> {
    map(
        alt((
            general_register,
            tag_no_case("ip"),
            tag_no_case("sp"),
            tag_no_case("fp"),
            tag_no_case("ret"),
        )),
        Atom::Register,
    )(input)
}

pub fn identifier(input: &str) -> IResult<&str, &str> {
    recognize(pair(
        alt((alpha1, tag("_"))),
        many0_count(alt((alphanumeric1, tag("_")))),
    ))(input)
}

pub fn variable(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("!")(input)?;
    map(identifier, Atom::Var)(input)
}

pub fn label(input: &str) -> IResult<&str, Atom> {
    let (input, _) = multispace0(input)?;
    let (input, label) = map(terminated(identifier, char(':')), Atom::Label)(input)?;
    let (input, _) = multispace0(input)?;
    Ok((input, label))
}

pub fn hex_literal(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("$")(input)?;
    map(hex_digit1, Atom::HexLiteral)(input)
}

pub fn address(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("&")(input)?;
    map(hex_digit1, Atom::Address)(input)
}

pub fn operator(input: &str) -> IResult<&str, Atom> {
    map(
        alt((
            map(char('+'), |_| Operator::Add),
            map(char('-'), |_| Operator::Sub),
            map(char('*'), |_| Operator::Mul),
        )),
        Atom::Operator,
    )(input)
}
