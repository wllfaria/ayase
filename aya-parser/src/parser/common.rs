use std::ops::Range;

use super::error::{bail, unexpected_eof, unexpected_token};
use super::Result;
use crate::ast::ByteOffset;
use crate::lexer::{Kind, Lexer, Token, TransposeRef};

pub fn expect<S: AsRef<str>>(
    expected: Kind,
    lexer: &mut Lexer,
    source: S,
    help: S,
    message: S,
) -> miette::Result<ByteOffset> {
    let current = lexer.next().transpose()?;
    let Some(current) = current else {
        unreachable!();
    };

    if current.kind != expected {
        return Err(bail(source.as_ref(), help.as_ref(), message.as_ref(), current.offset()));
    }

    Ok(current.offset())
}

pub fn expect_fail<S: AsRef<str>>(expected: Kind, lexer: &mut Lexer, source: S) -> miette::Result<bool> {
    let Ok(Some(next)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    if next.kind != expected {
        return unexpected_token(source.as_ref(), next);
    }

    lexer.next().transpose()?;
    Ok(false)
}

pub fn parse_identifier<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::Ident, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_register<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<ByteOffset> {
    let offset = parse_identifier(source.as_ref(), lexer, "", "")?;
    let name = &source.as_ref()[Range::<usize>::from(offset)];
    match name {
        "acc" | "ip" | "r1" | "r2" | "r3" | "r4" | "r5" | "r6" | "r7" | "r8" | "sp" | "fp" => Ok(offset),
        _ => unexpected_token(source.as_ref(), &Token::from_ident(name, offset.start, offset.end)),
    }
}

pub fn parse_hex_lit<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::HexNumber, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_string<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::String, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_variable<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::Bang, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    expect(Kind::Ident, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_keyword<S: AsRef<str>>(source: S, lexer: &mut Lexer, expected: Kind) -> Result<ByteOffset> {
    let next = lexer.next().transpose()?;
    let Some(next) = next else {
        return unexpected_eof(source.as_ref(), "invalid eof");
    };

    if next.kind != expected {
        return unexpected_token(source.as_ref(), &next);
    }

    Ok(next.offset())
}
