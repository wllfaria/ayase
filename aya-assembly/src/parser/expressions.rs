use miette::Result;

use super::ast::{Operator, Statement};
use super::common::{expect, parse_hex_lit, parse_register, parse_variable, peek};
use super::error::{HEX_LIT_HELP, HEX_LIT_MSG};
use crate::lexer::{Kind, Lexer};
use crate::utils::unexpected_token;

mod precedences {
    use miette::Result;

    use crate::parser::ast::Operator;

    pub const BASE: u8 = 0;
    pub const ADD: u8 = 1;
    pub const MUL: u8 = 2;

    pub fn from_operator(operator: Operator) -> Result<u8> {
        match operator {
            Operator::Add => Ok(ADD),
            Operator::Sub => Ok(ADD),
            Operator::Mul => Ok(MUL),
        }
    }
}

pub fn parse_literal_expr<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<Statement> {
    expect(Kind::LBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    let value = parse_expr(source.as_ref(), lexer, precedences::BASE)?;
    expect(Kind::RBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    Ok(value)
}

pub fn parse_address_expr<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<Statement> {
    expect(Kind::Ampersand, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    expect(Kind::LBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;

    let value = parse_expr(source.as_ref(), lexer, precedences::BASE)?;

    expect(Kind::RBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    Ok(Statement::Address(Box::new(value)))
}

fn parse_expr<S: AsRef<str>>(source: S, lexer: &mut Lexer, precedence: u8) -> Result<Statement> {
    let token = peek(source.as_ref(), lexer)?;
    let mut lhs = match token.kind {
        Kind::LParen => {
            lexer.next().transpose()?;
            let value = parse_expr(source.as_ref(), lexer, precedences::BASE)?;
            expect(
                Kind::RParen,
                lexer,
                source.as_ref(),
                "you likely forgot a closing parenthesis `)` [RIGHT_PAREN]",
                "[SYNTAX_ERROR]: unterminated expression group",
            )?;
            value
        }
        Kind::HexNumber => Statement::HexLiteral(parse_hex_lit(source.as_ref(), lexer, HEX_LIT_HELP, HEX_LIT_MSG)?),
        Kind::Ident => Statement::Register(parse_register(source.as_ref(), lexer)?),
        Kind::Bang => Statement::Var(parse_variable(
            source.as_ref(),
            lexer,
            "variable name must be a valid identifier",
            "[SYNTAX_ERROR]: invalid variable name",
        )?),
        _ => unexpected_token(source.as_ref(), &token)?,
    };

    loop {
        let token = peek(source.as_ref(), lexer)?;
        match token.kind {
            Kind::RParen => break,
            Kind::RBracket => break,
            kind if !kind.is_operator() => unexpected_token(source.as_ref(), &token)?,
            _ => {}
        }

        let operator = peek(source.as_ref(), lexer)?;
        let operator = Operator::try_from(operator)?;
        let operator_precedence = precedences::from_operator(operator)?;

        if operator_precedence < precedence {
            break;
        }

        lexer.next().transpose()?;

        let rhs = parse_expr(source.as_ref(), lexer, operator_precedence)?;
        lhs = Statement::BinaryOp {
            lhs: Box::new(lhs),
            operator,
            rhs: Box::new(rhs),
        }
    }

    Ok(lhs)
}
