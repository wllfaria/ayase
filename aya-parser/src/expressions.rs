use nom::branch::alt;
use nom::bytes::complete::tag;
use nom::character::complete::space0;
use nom::IResult;

use crate::common::{get_precedence, hex_literal, is_terminator, operator, peek, precedences, register, variable};
use crate::types::Atom;

fn expr_group(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("(")(input)?;
    let (input, _) = space0(input)?;

    let (input, expr) = parse_expr(input, precedences::BASE)?;
    let (input, _) = space0(input)?;

    let (input, _) = tag(")")(input)?;
    let (input, _) = space0(input)?;

    Ok((input, expr))
}

fn expr_item(input: &str) -> IResult<&str, Atom> {
    alt((hex_literal, register, variable, expr_group))(input)
}

fn parse_expr(input: &str, min_precedence: u8) -> IResult<&str, Atom> {
    let (input, mut lhs) = expr_item(input)?;
    let (input, _) = space0(input)?;

    let mut full_input = input;

    loop {
        let (_, next) = peek(full_input)?;
        if is_terminator(next) {
            break;
        }

        let (input, operator) = operator(full_input)?;
        let (input, _) = space0(input)?;

        let Atom::Operator(operator) = operator else {
            unreachable!();
        };
        let precedence = get_precedence(operator);

        if precedence <= min_precedence {
            break;
        }

        let (input, rhs) = parse_expr(input, precedence)?;
        let (input, _) = space0(input)?;

        full_input = input;

        lhs = Atom::BinaryOp {
            lhs: Box::new(lhs),
            operator,
            rhs: Box::new(rhs),
        }
    }

    Ok((full_input, lhs))
}

pub fn square_bracketed_expr(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("[")(input)?;
    let (input, _) = space0(input)?;

    let (input, expr) = parse_expr(input, precedences::BASE)?;
    let (input, _) = space0(input)?;

    let (input, _) = tag("]")(input)?;
    let (input, _) = space0(input)?;

    Ok((input, expr))
}
