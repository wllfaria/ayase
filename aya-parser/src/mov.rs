use nom::branch::alt;
use nom::bytes::complete::tag_no_case;
use nom::character::complete::{char, space0, space1};
use nom::sequence::preceded;
use nom::IResult;

use crate::common::{address, hex_literal, register};
use crate::expressions::square_bracketed_expr;
use crate::types::Instruction;

fn mov_lit_reg(input: &str) -> IResult<&str, Instruction> {
    let (input, _) = tag_no_case("mov")(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = alt((hex_literal, square_bracketed_expr))(input)?;
    let (input, _) = space0(input)?;

    Ok((input, Instruction::MovLitReg(lhs, rhs)))
}

fn mov_reg_reg(input: &str) -> IResult<&str, Instruction> {
    let (input, _) = tag_no_case("mov")(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = register(input)?;
    let (input, _) = space0(input)?;

    Ok((input, Instruction::MovRegReg(lhs, rhs)))
}

fn mov_reg_mem(input: &str) -> IResult<&str, Instruction> {
    let (input, _) = tag_no_case("mov")(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = alt((address, preceded(char('&'), square_bracketed_expr)))(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = register(input)?;
    let (input, _) = space0(input)?;

    Ok((input, Instruction::MovRegMem(lhs, rhs)))
}

fn mov_mem_reg(input: &str) -> IResult<&str, Instruction> {
    let (input, _) = tag_no_case("mov")(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = alt((address, preceded(char('&'), square_bracketed_expr)))(input)?;
    let (input, _) = space0(input)?;

    Ok((input, Instruction::MovMemReg(lhs, rhs)))
}

fn mov_lit_mem(input: &str) -> IResult<&str, Instruction> {
    let (input, _) = tag_no_case("mov")(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = alt((address, preceded(char('&'), square_bracketed_expr)))(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = alt((hex_literal, square_bracketed_expr))(input)?;
    let (input, _) = space0(input)?;

    Ok((input, Instruction::MovLitMem(lhs, rhs)))
}

fn mov_reg_ptr_reg(input: &str) -> IResult<&str, Instruction> {
    let (input, _) = tag_no_case("mov")(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = preceded(char('&'), register)(input)?;
    let (input, _) = space0(input)?;

    Ok((input, Instruction::MovRegPtrReg(lhs, rhs)))
}

pub fn parse_mov(input: &str) -> IResult<&str, Instruction> {
    alt((
        mov_lit_reg,
        mov_reg_reg,
        mov_reg_mem,
        mov_mem_reg,
        mov_lit_mem,
        mov_reg_ptr_reg,
    ))(input)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{Atom, Operator};

    fn run_instruction(input: &str) {
        let (input, instruction) = parse_mov(input).unwrap();
        insta::assert_debug_snapshot!(instruction);
        assert!(input.is_empty());
    }

    #[test]
    fn test_mov_lit_reg() {
        let input = "mov r1, $c0d3";
        run_instruction(input);
    }

    #[test]
    fn test_mov_reg_reg() {
        let input = "mov r1, r2";
        run_instruction(input);
    }

    #[test]
    fn test_mov_reg_mem() {
        let input = "mov &c0d3, r2";
        run_instruction(input);
    }

    #[test]
    fn test_mov_mem_reg() {
        let input = "mov r4, &c0d3";
        run_instruction(input);
    }

    #[test]
    fn test_mov_lit_mem() {
        let input = "mov &c0d3, $ffe3";
        run_instruction(input);
    }

    #[test]
    fn test_mov_reg_ptr_reg() {
        let input = "mov r3, &r4";
        run_instruction(input);
    }

    #[test]
    fn parse_complex_mov_reg_lit() {
        let input = "mov r1, [$42 + !loc - ($05 * ($31 + !var) - $07)]";
        run_instruction(input);
    }
}
