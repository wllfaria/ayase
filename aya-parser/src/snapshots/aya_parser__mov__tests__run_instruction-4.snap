---
source: aya-parser/src/mov.rs
expression: instruction
---
MovRegPtrReg(
    Register(
        "r3",
    ),
    Register(
        "r4",
    ),
)
